### ISSUE TITLE
Keystone and the future

### ISSUE THREAD

**Issue by autoboxer** [keystonejs/keystone#4551]
Activity on all channels related to this repo ( Slack, Gitter, and here ) seem to have dropped off a while ago, picked up for a brief moment a few months ago, then dropped off again.  There are many people in the community who are reliant on this repo for projects they're working on ( myself included, I just spent almost 4 years building a 30,000+ line codebase for a local adoption agency on Keystone ), and we need to know how to proceed with our work related to this repo.  If this is still active, please let us know and start communicating, there are many people who need to know.

**Comment by ttsirkia** [comment]
The situation is very sad. I think the recent activity was just to fix some critical security issues. The version 4 has been in beta phase so long without any real development that it is hard to belive that the situation would change. The sudden stop of the development has also stopped the community because there is no point to make pull requests as they are not merged and nobody knows how the project will continue.

If the Thinkmill company who has been involved with the project has proceeded with the development as mentioned here earlier, they must have a private repository because nothing is visible here.

Keystone would have become the best JavaScript based CMS and framework if it had been managed better. At least in my opinion, the framework is very well done and the admin UI is one of the key features that other similar projects do not have.

I really hope that the project will continue and the stable version 4 is being published soon. But I'm not very optimistic. Are there any maintained forks?

**Comment by autoboxer** [comment]
I agree @ttsirkia.  One of the reasons I was asking was in case it had been abandoned; I don't have much time, but I have a lot invested in Keystone, as I'm sure many people do.  I'd be happy to build out a new organization and fork the code, as well as help understand and document what exists now.  I just wish that they would say something instead of disappearing and not communicating.  This has the potential to be the go to JavaScript CMS, and I'm hoping there are enough people out there willing to help push it to what it could become.

*To the general community*: If this doesn't get any attention from core contributors, please respond here if you're interested in helping to maintain the project.

**Comment by Elemino** [comment]
I would definitely like to help this project stay alive! Count me in! I am very much invested and believe there's nothing like it within the open source community. I noticed the documentation links are broken.

**Comment by christroutner** [comment]
So food for thought: if this project was forked and managed by another
organization, who will pay for it? who would manage it? What's in it for
them?

I think it's wonderful what ThinkMill has created. I use KeystoneJS all the
time, but I don't use any of the more modern changes in v4 beta. I've been
pretty content to use it as is.

It's always better if a project can be actively managed and improving
through pull requests. But what's the incentive for someone to take on a
large project like KeystoneJS? It's got to be more than desire. There has
to be some financial incentive for an organization to curate the code base,
in order for the project to live on long-term.

**Comment by autoboxer** [comment]
I agree @christroutner about how to handle the codebase through pull requests, although there needs to be a core team who know the codebase well enough to approve or deny changes.  As far as cost, I'd be happy to eat it to get things going (both the site hosting and the github organization), and would make the core contributors admin along with me so we don't get into a situation where an individual has too much control.  I'm busy through the end of the month, and feel a few weeks time to let the core team speak up if this repo is not in fact dead would be appropriate.  In addition, if this seems like a viable option, I'd like to try to contact Jed and Max first, as well as send a general message Thinkmill to see what's going on.  If we don't hear anything by then, I'll send out a direct message to everyone on this thread and any who want to be involved can talk about how best to proceed.  I personally think there needs to be a lull period where we all read through the codebase, compare notes, and fully understand how everything works together before anything can be reliably done, as maintaining the integrity of the codebase should be the highest priority.  Based on what I've seen, there needs to be an early push for stabilizing the codebase and determining what is needed to consider cutting a stable release.  Things that I see missing are:

clear documentation for the 4.x branch
easily accessibly documentation for upgrading versions
more robust code commenting
strategy around pull requests and the review process
more complete unit testing
security analysis and vulnerability fixing
analyzing and upgrading dependencies

In addition, there are fixes that were made it into the 0.3.x branch that have not been ported to the 4.x branch, which will need to be handled.  The original team has a ton of documentation and roadmapping here, on their site, in the Google Group, on their Slack channel, in Gitter, and they have a voting app somewhere to vote on the next features to include, which could be leveraged.

As of now the people I'll be contacting are: @ttsirkia, @Elemino, @christroutner.  If any of you don't want to be included for any reason, just let me know and I'll make sure you're not bothered by the communication at the end of the month.

**Comment by ttsirkia** [comment]
I would be extremely happy to see the beta tag removed and version 4.0 published. The feature set is enough for my purposes.

I'm still running 0.3.x and the dependency libraries (Mongoose and others) and Node are nowadays really old. Apparently there are some bugs in the current beta release. I haven't started the migration process to the new version yet because I should be sure that it is a good decision.

if the framework is no longer maintained, it might be good idea to re-invent the whole website with another framework. That is, of course, a huge task and therefore I really hope that the Keystone project would continue in some way.

**Comment by larskarbo** [comment]
I have a few projects that use keystone, and would very much like to help keeping this project alive! I have some spare time coming up and won't mind reading up on the code base

**Comment by Elemino** [comment]
I believe Keystone can be very successful financially in one way or another. I'll go so far as to say that in the long run, it seems plausible to me that Keystone might be able to provide an ecosystem or a service that will allow non-developers to build and deploy through an admin UI along with its use of modularity. Wherever this project may roam, whether if it stays a developer tool or not,  I would be willing to contribute my time and efforts in any way possible. @utoboxer Thanks for stepping up and I'm looking forward to your correspondence. 

I'll try to do my best and go over the code base to familiarize myself with the ins and outs in the days to come.

**Comment by quitequinn** [comment]
I'm not confident I have the time or skill to properly contribute to a fork (if it happens) but I am invested in keystone and would like to contribute if possible.

**Comment by bflopez** [comment]
I am on the fence of getting my company to commit to KeystoneJS. It really felt like I had found the right tool for what we needed. I was already considering forking the project and maintaining my own version specific for my company. I am willing to help.

**Comment by ttsirkia** [comment]
Of course it is the best alternative that @JedWatson continues the development. And that would be a natural choice because Keystone seems to still be one of the key components of the ThinkMill company. Hopefully we hear something from them soon.

**Comment by autoboxer** [comment]
I couldn't agree more @ttsirkia, I'd love nothing more than for @JedWatson, @mxstbr, or someone else from the core team to come here and let us know the project is in good standing.  If that's the case, we can all happily go back to contributing as maintaining a project of this size could be a considerable commitment.  Most of the conversation above is centered solely around the case where no one responds.

**Comment by dlombard** [comment]
@autoboxer Count me in. I have not officially contributed to the KeystoneJS but I've used it and ready to spend sometime to continue the development

**Comment by hreimer** [comment]
I am also willing to help with various fixes, small tasks, though I'm not yet experienced enough to be able to contribute actively to a fork/this repository or oversee if a PR is good to be included..

Since I started using Keystone last summer for a project I always hoped for the project to continue, however since there were virtually no new features or PRs included it became kind of necessary to develop custom functions - on one of those occasions I messaged @mxstbr on Twitter, and regarding a possible involvement in the project he stated that he was not part of it anymore and unfortunately couldn't help me. (at least it seemed to be the case on Oct 5, 2017, hope it's not a problem writing this here)

**Comment by yrmichael** [comment]
While it's disappointing to see keystone neglected, I'm yet to find anything I can't do and can't live without using keystone. i'll be hoping it has a renaissance this year, but even if it doesn't, i'll still be recommending it and using it myself, though likely forked.

**Comment by MichaelZaporozhets** [comment]
While it's disappointing to see keystone neglected, I'm yet to find anything I can't do and can't live without using keystone. i'll be hoping it has a renaissance this year, but even if it doesn't, i'll still be recommending it and using it, albeit forked, myself.

**Comment by gionatan-lombardi** [comment]
I'm trying to migrate from Wordpress to Keystone in order have everything in the Node environment, but I need some important features to really do that in all my projects. Some of these, like nested lists or custom fields, are totally still or the PRs are blocked. I've started to read the codebase and'll be happy to actively contribute, count on me too.

**Comment by marconucara** [comment]
Excellent initiative. I have not contributed to the KeystoneJS, but I would like to.

**Comment by internetErik** [comment]
Would love to see Jed and other core contributors continue to lead this project, so I hope to hear from them soon. If we do hear back, maybe the core team can somehow take advantage of all of this interest in people contributing and help a bunch of us get set up do to so.

@autoboxer  If there is a fork, I am interested in contributing.

**Comment by autoboxer** [comment]
it's so nice to see so many people invested, and I definitely feel more comfortable in my Keystone-dependent work seeing so many people respond so quickly.  I'm keeping a running list in a comment above of everyone who mentioned the possibility of contributing or added to this conversation in a serious way.  In an early effort to get the team involved, @JedWatson and @mxstbr were mentioned in comments, and if anyone else knows others who would be worth reaching out to, please post those names here so this gains visibility with the core team.  I'd like to make sure we do our due diligence first instead of making assumptions as to the project state as there are still avenues to check.  As I said before, around the end of the month I'll be reaching out to Thinkmill if we don't hear anything.

**Comment by mxstbr** [comment]
While I no longer work at Thinkmill and thusly can't give any official response regarding the current maintenance of Keystone, I know that Thinkmill built all of their projects with Keystone at the core and I highly highly doubt they switched away or are using something else.

Most likely, Keystone just works as-is for them, which is why commit activity has slowed down outside of security patches. (on top of them most likely being super busy with commercial work)

**Comment by igorsamado** [comment]
I would like to contribute somehow... KeystoneJS cacth me when I was looking for a good NodeJS CMS for my personal site. I was developing a local  version of my site with keystonejs on my notebook.. I was reading a little the codebase...

**Comment by JedWatson** [comment]
Thanks for opening this issue @autoboxer, and everyone here for the attitudes expressed. It's really heartening to hear that people find so much value in something we've created, and are so interested in seeing it continue.

I'm going to try and address as many of the points as I can, and help move things forward.

@chrisroutner expressed an insightful point:

> But what's the incentive for someone to take on a large project like KeystoneJS? It's got to be more than desire. There has to be some financial incentive for an organisation to curate the code base, in order for the project to live on long-term.

We (speaking as a co-founder of Thinkmill) have used Keystone in most of our projects over the last few years, and that's helped us continue to develop and maintain it; even including periods where we've employed people (like @mxstbr - miss you buddy!) full time to focus on it.

However, it's also been a labour of love. I (and others) have poured countless hours into it _just because I wanted it to exist_. There have been evenings I've worked until early morning coding, triaging issues and PRs, or skyping with contributors around the world. But over the course of years, and with a new child, that ended up not being sustainable.

Meanwhile, many of the things we do at Thinkmill evolved and Keystone as-is has been both enough (in terms of features and maturity) and not enough (in terms of bigger features that have proven hard to build into the current v4 architecture). Because of this, unintentionally, we've ended up doing a lot of work _around_ Keystone instead of being able to contribute _into_ it; which, as a business that's grown from 5 to 20 people over that time, has made it hard to give the project the attention we've wanted to.

What we want to do is not just deliver on a solid release of keystone 4, but also be looking at what we want to make 5 6 and beyond, so it remains the go-to node CMS.

We've been trying to work out what to do about this, and what the future looks like, and I've erred on the side of not saying anything (while having long internal discussions about it) because I don't know what to say yet; I'm sorry to everyone it happened like that, because I know it hasn't been good for the community.

So here's what I need to say:

We haven't actually abandoned Keystone, just done a terrible job of communicating what's going on and empowering others to be involved. So we're fixing that, starting today.

As has been pointed out, v4 should have been out of "beta" ages ago and, functionally, has been for a year. It needs to be cleaned up and published so we can move forward.

We'd like to bring in new admins and maintainers from the community to help organise and do the work of finalising v4, and continue maintaining it into the future. Basically, what everyone's talking about here - but without the need to fork, and with our support and involvement.

I'm happy to say that @noviny is stepping up as community manager on behalf of Thinkmill, so I'm going to hand over to him to outline the plan from here.

**Comment by ttsirkia** [comment]
Great, thanks @JedWatson for the answer!

**Comment by ttsirkia** [comment]
I think this is the best possible situation if Thinkmill has both commercial need and desire to develop it further.

As seen in this thread, there many developers who are willing to contribute to this project, which makes the framework even better for all. Of course this requires some management to lead the project and go through issues and pull requests.

For me this was clear indication that updating from 0.3 to 4 can be started and in some point, the repository with the current features but up-to-date dependencies are available. After the version 4 is officially released, community can again start to contribute as we know that there are not any major changes coming soon which would break all PRs.

**Comment by autoboxer** [comment]
@JedWatson and @mxstbr, thank you both so much for taking the time to reach out here and add clarity to the conversation.  @JedWatson, the news you gave is, in my eyes, the best possible for the community, and I'm sure everyone here is thrilled that Keystone is in fact still active.  Hopefully this thread shows the willingness of the community to help with the next steps, and @Noviny, as you transition into community manager, please let us know what you need as I'm sure many here will be willing to help along the way.

**Comment by dlombard** [comment]
@JedWatson Thanks a lot for your input! I am happy to hear the Keystone is still actively used and that Thinkmill will keep putting effort to improve Keystone

` Keystone as-is has been ... not enough (in terms of bigger features that have proven hard to build into the current v4 architecture).`

If I may @Noviny  & @JedWatson , what kind of features are we talking about? Is a partial re-architecture of Keystone is considered for post 4 releases?

**Comment by marconucara** [comment]
I think that an updated roadmap would be great now. For me, the missing key features are nested list, custom fields, subpages/pagetree and multi languages but I'm curious about your ideas about that.

**Comment by JedWatson** [comment]
Just a quick note to say I updated the title of this issue to something I hope now better reflects the outcome. Happy to leave it open until we've actually got some momentum towards the steps @autoboxer, @Noviny and myself are discussing though.

**Comment by tvalentius** [comment]
I've been using KeystoneJS for Javascript community in Bali ( http://balijs.com/ ) and i think Keystone is one of the most mature, robust, and easy-to-modify NodeJS based CMS.

Please count me in, i am really excited to contribute more on KeystoneJS :dagger:

**Comment by MichaelZaporozhets** [comment]
Please count me in also. I lead on and maintain a variety of projects large and small built atop keystone, and I’d love to help grow it.

**Comment by CAYdenberg** [comment]
I would like to help if possible (I enjoy writing unit tests for React components). I've kept my distance only because my few small PRs were generally ignored. I really believe in this project and think with a few bells and whistles and a robust plugin interface Keystone could easily blow away more "mature CMSes" (not naming any!), esp. on the "developer happiness" scale.

**Comment by n-devr** [comment]
This is great news!  Glad to hear there is still interest in maintaining and enhancing KeystoneJS!

I have been working with @autoboxer on his adoption agency project and leveraging Keystone for a few of my personal projects as well.  I'd like to see the project and the community continue to flourish and I'd be happy to be a part of that effort.

**Comment by autoboxer** [comment]
I'm looking forward to helping @JedWatson and @Noviny, and I've updated my profile to make my email public so we can begin the dialogue.  From the looks of this thread, you'll have no shortage of contributors to help get Keystone 4 and beyond developed, documented, and released.  I live in Boston, MA  which is Eastern Standard Time ( GMT-5 ), but can be flexible around communication times to make it more convenient to your team.

**Comment by Elemino** [comment]
I have experience with continuous integration and today I'm trying to figure out how to test keystone properly. I'll be using jed's keystone test project as a reference and document the process. My goal is to test payment integrations first and foremost for my online shop example. I'll document the process and see and it can lead to contributions that way. I see myself helping with the dev opsy side of things.

**Comment by danielmahon** [comment]
🎉  Woot! 🎉  

I'm very excited to hear the good news and would love to get more involved when the plan gets laid out. I use Keystone on a handful of commercial sites and apps that I manage. Most are rather low traffic so keeping them in line with the latest Keystone isn't really paramount. But I also build out a lot of new projects and proof-of-concepts that utilize Keystone primarily as their database interface or API server. Being able to easily customize models and not have to mess too much with the UI is a HUGE benefit for me. I work mostly on my own and, because of Keystone, am able to maintain projects that should be way beyond the scope of one person.

I think breaking up the scope of Keystone development will help the community more easily contribute to its growth. I know this has been in the pipeline for a while now, with the separation of certain modules and concepts like plugins, etc. Right now it feels tightly coupled and its rather difficult I think to wrap your head around everything Keystone is doing behind the scenes if you're not familiar with it. Breaking things up like Types and such into their own repos might help focus effort where it's needed and allow contributors to work with more manageable chunks without interfering with core efforts.

I for one have spent a lot of time "hacking" on the CloudinaryImage Type and parts of it still confuse me 😉  I think taking a step back and simplifying the core Keystone solution will help it's own ecosystem grow and become more sustainable.

I couldn't imagine using anything else at the moment, even as it stands it's an amazing tool and I'll gladly do what I can to help it thrive.

**Comment by morenoh149** [comment]
⚔️  still can't find a Django or Rails esque solution for node. Monolithic ftw until you outgrow it. Loopback upgrades are not good and their business model is to sell support not have a robust framework. Sails is bleh. http://www.nodaljs.com/ is alright, if we could extend the keystone ORM to support RDMSs and nosql (elasticsearch for example) that would be a game changer.

**Comment by JedWatson** [comment]
The amount of support here is quite overwhelming - thanks everyone. I hope we can make the most of this new phase and live up to the promise 😄 

With so many people interested in helping (and in various different areas too) I've been thinking about how best to start and it's probably best to just lay out some ideas, see what resonates with others.

First up, I think part of the problem we (the original team) have been facing is that Keystone development feels _hard_. The way the tests are set up makes it painful (at least, for me) to move quickly through development. The state management in the Admin UI has been difficult to work with (again, for me) since Redux was introduced. And because Keystone is so flexible, it's hard to keep all the requirements and features that could interact with each other in your head at one time - this is where something like the test project comes in.

So if we're all going to work on it together, I think we should start by solving these problems and making it easier for everyone to work together and contribute (whether a lot or just a little, as we each can)

My rough ideas are:

- Start establishing a set of "requirements" or features that Keystone meets, ideally as part of the documentation. Not having all the behaviour specified up front has become a burden over time. Note I'm not talking about a roadmap or future wish list; this is to make it easier to understand what's there and how it's intended to work so we can develop with confidence
- Pull the plug on the old tech that's currently supported; move to a really modern version of node as the minimum requirement (say, 8) so we can use modern tooling
- Focus on the [test project](https://github.com/keystonejs/keystone-test-project) (for the Admin UI and end-to-end testing) and script it with [cypress](https://www.cypress.io) for both local testing and CI
- Get proper preview deployments of the test project set up as part of the CI (I have added this recently to (react-select)[https://github.com/JedWatson/react-select/] using [netlify](https://github.com/JedWatson/react-select/) and it's made such a huge difference)

We actually do have a couple of things we really should do in the short term as well, including pushing out the changes that are required to upgrade `multer`. @noviny will circle back with a more comprehensive list, and we can go from there.

There's also the the current PR and issue backlog - would be great if we can get some people to help work through that. It might even be a good idea to simply close a lot of the old stuff and start again with it so we can more effectively use Github to collaborate. I'm open to everyone's feedback on this.

Bigger things that everyone wants - like nested lists, custom fields, etc - are really honestly blocked behind the "boring" stuff I've listed above and if everyone can actually help us clean this project up so it's in a good place for v4 and easier to work on / contribute to, then I'm really excited about the potential stuff we could do after.

---

To answer @dlombard's question, and give some insight into what we think the future is post-4.0:

From the research and prototyping we've done over the last few months, we are keen to change the architecture of Keystone pretty significantly to separate it into different concerns and make each easier to integrate with other parts of the node / JavaScript ecosystem. "Break down the monolith" if you like. So the server backend, Admin UI, and core List Schema + db abstraction can all be dealt with independently.

We're also pretty bullish on GraphQL and love what tools like [graph.cool](http://www.graph.cool) and Apollo give you in terms of developer productivity. Those systems are still missing some of what Keystone gives you - especially the Admin UI and custom, complex field types - but there's overlap with things like build processes and web servers that we're starting to see.

We also think the next generation will move away from being explicitly tied to MongoDB via Mongoose.

Finally, the user-facing things: finally making the Admin UI extensible, with custom field interfaces / routes / components etc; implementing a built-in publishing workflow and better editor for content management is pretty high on my (personal) list too.

**Comment by jfmario** [comment]
I've recently worked on a webapp using Keystone.js, and since it was the closest thing to Django (the first framework I worked with) I fell in love it. I'm more junior, but I'd also like to be involved in helping if I can.

**Comment by dlombard** [comment]
Thanks alot @JedWatson for your feedback.

I, also, agree with the rough ideas you've listed as they are aimed at solving the core of the product (the "boring" stuff as you say 😃 ) before dealing with functionalities.

I've been reading all the comments here and I'd like to add some food for thought on the table if I may.

I think that before listing the requirements and features, it would be good to determine the problem(s) that KeystoneJS is meant to resolve (What you want Keystone to be ultimately). Right now, Keystone is highly flexible; I could think of it as the swiss knife of web development as it allow to create APIs, web apps and CMSs. To me it makes it an amazing product. However, that flexibility brings its load of difficulties and may result in a pretty long list of requirements and features. At the very least, defining what KeystoneJS should be (or what problem it solves) will help steering future conversations and development in the right direction. It will also serve as a reference point, ensuring that reqs and features will make KeystoneJS more of what it should be.

Then, as you mention we should start establishing a set of "requirements" or features that Keystone meets. Having a full list a features and requirements is one of the best ways to bring a project to completion

I love the idea of "breaking the monolith" that you brought up. It will help to deal with the different pieces of the puzzle independently.

> We also think the next generation will move away from being explicitly tied to MongoDB via Mongoose.

Are we talking about supporting multiple databases (database agnostic) or developing a custom Keystone model like loopback does?

Thanks!

**Comment by jstockwin** [comment]
@Noviny and @JedWatson - Glad to see things moving again! As you both know I was quite heavily involved in setting up the current E2E tests.

I understand they are quite painful at the moment (specifically because they're split across two repos) and can see why you want to change them. That said, I might be able to give you some insight into why they are set up as they are, so that you can keep this in mind while making any changes. I feel there is a reasonable amount of knowledge from discussions between myself and @webteckie while creating the E2E tests that need to be communicated to you.

@Noviny we had a skype call set up at one point which didn't happen. The offer is still there if you want it.

**Comment by gautamsi** [comment]
Very glad to see it is not dead yet. I lost hope sometime ago and started forking to TypeScript (I am big fan of TypeScript if you look at my github profile.) Ended up with https://github.com/gautamsi/keystone-ts (still no build instruction, will add soon)

My idea was to go from es5 classes to ES6 classes with TypeScript and use other recent features introduced by nodejs

I completed the current port **last night** and tested [sydjs-ts](https://github.com/gautamsi/sydjs-ts) (which is just a TypeScript port of [sydjs-site](https://github.com/JedWatson/sydjs-site)) works fine with some React warnings. Plan to use new features like async/await soon.

work done so far:
1. converted all es5 classes to TypeScript classes compiling to ES6. 
   1.1 merged some of the code from /lib/core and /lib/list to KeyStone and List class.
2. upgraded React to 16.2 with other dependencies (took longest being new to react)
   2.1 All React code is also in Typescript (.tsx), extended from Component and converted to pure component using TypeScript.
3. Converted FieldType to Class and have other Types inherit from it (it was difficult due to super_ being called at the end which is not possible in ES6)
4. Added properties (List/View/utils etc.) on Keystone instance to give full compatibility with existing code using keystone.

Plans: 
1. Complete unit tests
2. Change List constructor API, use Typed FieldType rather than POJO in list.add method.
3. Streamline Mongoose Schema to make it easier in extending field types and adding field types
4. introduce plugin architecture for admin and hook for use plugins in templates. 
5. integrate some of the pull requests from this repo. like one with role based access
6. move away from bable and dynamic creation of FieldTypes module with browserify. 
    6.1 plan to pre build this as part of build process.
7. integrate publishing workflow (in work branch in keystone)
It would be wonderful to see ES6/TypeScript along the way for 4.0 release. 

Would be glad if @Noviny or @JedWatson  want to integrate my TypeScript work here.

**Comment by autoboxer** [comment]
@Noviny and @JedWatson, you most likely have a lot to do to prepare for any upcoming coding, so please let me know if there's anything that I, or we as a community, could do to help.  I think it's important you have the time to come up with the simplest possible roadmap to a 4.0 release, and are able to focus on keeping things moving, so please utilize us as much as you need.  The following are less glamorous tasks that I, and I'm sure many others, would be happy to help work on/facilitate in the community/oversee:

1. **cleaning up/organizing issues from GitHub or any other source** ( whether it's reviewing/testing/commenting on open issues, or consolidating information for the team ).
2. **cleaning up/organizing the GitHub area itself** ( creating/updating labels, creating/updating templates ).
3. **reaching out and cataloguing new ideas and discussion from the community through the many platforms that host Keystone discussions, and consolidating the information in an easy place for the team to access.**
4. **writing documentation.**

Please let us know, there are clearly many here who are eager to help, but ultimately we need to follow the direction of the team and all be aware enough of what's planned to help move things towards a common goal.

**Comment by christroutner** [comment]
@JedWatson since this thread has the scope of 'how to start again', here are my viewpoints as someone who has used KeystoneJS for dozens of projects over the last two years. @Noviny, @autoboxer, maybe some of the following can be incorporated into plans to leverage community help and generate a new road map. I think a lot of the coupling that I talk about below is preventing wider community engagement.

If my criticisms below are harsh, it's only because I love this project and I have a clear vision of what it could be. I think adoption of this software could be much greater if some of the below problems were solved.

#### Separating the back end from front end:
I don't use any of the React features of KeystoneJS. I use it almost exclusively as an NPM library that has a convenient Admin UI. I think there are a lot of users like me. I never agreed with the direction the project took to adopt React. I think KeystoneJS should be as front-end agnostic as possible. And it is, to a certain extent. It's nice to have a front-end Admin UI (which must use one of the existing front end frameworks, I get that). But going forward, care should be taken not to lock users into any particular font end framework. 

To this end, I love the work that has been done with the Admin API. I use it extensively in my development, even though it's still pretty limited.

With regards to new features like nested lists and other new field types, their behavior and functionality on the back end should be uncoupled from the front end. This is especially important for tests. Just because something like a nested list doesn't render in the Admin UI front end, doesn't mean it doesn't work. Likewise, just because things like text arrays display well on the front end, doesn't mean their actual implementation on the back end isn't buggy (it is).

#### Testing
If decoupling front end from back end stuff is taken seriously, then it makes testing much easier. I almost never run tests in my Keystone installs because my experience is that they are flaky and unreliable. End-to-end testing doesn't provide any value to me, because I don't rely on the front end. Unit tests are great, but as an end user, I don't really care about them, except to see if the build is broken. As an end user, my primary concern is integration tests, which validate the environment and tell me if the installation is working correctly.

It would be great to see more focus on integration tests. It would be great to see integration level tests that tell me the following:
* confirm that ports are open
* confirmation that API end points are returning expected data and are functioning correctly
* confirmation that user-model CRUD operations and permissions are functioning correctly
* confirmation that the database and Keystone JS are communicating correctly.
* confirmation that interaction between any webservices (like Mailgun, Cloudinary, S3 storage, etc) are working correctly.

#### Dividing the Monolith - Templates and Plugins
You mentioned that KeystoneJS has grown to be fairly monolithic. That's my experience too. It would be great to see KeystoneJS slimmed down to a sleek library that is really just Express on steroids. Additional features like webservice hooks or even the Admin UI could be added on as additional, separate libraries.

When I was developing [ConnextCMS](http://connextcms.com), I very quickly ran into this issue, and KeystoneJS had not (still has not?) developed a system for site templates (for website content) or plugins (for additional APIs). I developed both for ConnextCMS and that code could easily be rolled into KeystoneJS.

* [This hack at the end of keystone.js](https://github.com/skagitpublishing/connextCMS/blob/b6abdf89b2a89c7f071f007de186d398b046eecb/keystone-files/keystone.js) allows me to load plugin and theme packages at run-time.

* [This hack to the routes/index.js file](https://github.com/skagitpublishing/connextCMS/blob/b6abdf89b2a89c7f071f007de186d398b046eecb/routes/index.js) loads new API end points at run-time.

* After implementing those hacks, I was then able to develop a [plugin template](https://github.com/skagitpublishing/plugin-template-connextcms) and a [theme template](https://github.com/skagitpublishing/site-template-connextcms) and eventually a [series of plugins](http://connextcms.com/page/plugins) to extend ConnextCMS features. All this code could easily be rolled into KeystoneJS, because it's all based on KeystoneJS.

#### Stop breaking the master branch
This is less of an issue in the last year, but when I first started playing around with KeystoneJS, the `master` branch seemed to be broken frequently. I started the habit of forking the master branch on days when I could confirm that installation worked correctly, as insurance against days when it didn't. I would also create repos with all the compiled NPM dependencies too, as reliably compiling those was also a frequent headache.

Seeing frivolous GitHub Issues from new users, stymied by a broken master branch or unclear node version dependencies, have been the norm. More effort needs to be taken to prevent this. I know from reading the Google mailing list and the GitHub issues that a lot of prospective developers and users are abandoning the use of KeystoneJS because of these continual problems.

**Comment by tvalentius** [comment]
Regarding React, i have to disagree with @christroutner , one of the reason for me (and i believe lot's of users too) interested to contribute on KeystoneJS, is because the project use and leverage React for the client-side code base.

It is a good idea to make an "agnostic" version of KeystoneJS though, but i am a big proponent of us going for React direction for the main/master branch of this project.

**Comment by marconucara** [comment]
I can't got @christroutner's point about React. It's used only for admin UI, users aren't force or encouraged to use it in their own frontend. Am I wrong?

**Comment by Elemino** [comment]
So far it seems to me that cypress.io covers most if not all of those integration tests that @christroutner mentioned. I can also see a development branch being created for front-end agnostic purposes while continuing leveraging react in the keystone. @JedWatson seemed really excited about graphql and incorporating modern tools and I would like to see that type of excitement prevailing! Maybe we can create different channels to debate different issues on gitter before this thread gets too long so everybody is free to speak up and share their thoughts. My goal is to take the pressure out of the core team while enabling free debate. @christroutner your comments are super helpful and valued and I am learning a lot!

**Comment by JedWatson** [comment]
Thanks for continuing to pitch in, everyone.

@jstockwin absolutely, you and @webteckie know the most about how things are. To be clear I don't mean any criticism of your work, just that technology and tools have moved on and I do believe we can set it up to be easier / more productive, and even better if you can help or advise.

I am plenty critical of my own work as well 😉 

@gautamsi great to hear you've made so much progress but to be completely honest I don't know TypeScript and don't expect to use it in the near future - @Noviny and I actually use Flow extensively day-to-day. I am a strong believer of the value of having a type system in JS but right now our focus needs to be on making Keystone simpler to contribute to and maintain, and needing to up-skill everyone in a new language doesn't seem like a good idea in the short term. Let's keep an eye on that in the future.

That said, upgrading the client-side codebase to modern ES8 (with babel) and being able to use new features with recent node versions is a high priority.

@christroutner thanks for the extensive write-up. I agree with a lot of your sentiments.

Keystone is, and I expect always will be, front-end agnostic. You can create server-side routes with Express, set up a REST api and build in something like browserify/webpack, or build your front-end independently in any language and wrap the whole thing in a GraphQL layer - we've done all of that today with v4.

If anything, I'd like it to be easier to drop in other modern tools like Next.js or Gatsby and have them easier to integrate without feeling like you need to "hack" them into your Keystone app. Same for any tooling like Webpack and Parcel, and front-end frameworks like Angular and Vue. As of today it's not that obvious how separate these things actually are, and we need to improve that going forward.

That said the Admin UI is and will continue (for at least a long time) to be powered by React.

re: testing, 100% that's been my experience.

Same with the master build. Often - in my experience - it's not been an issue that Keystone is actually _broken_, just that the tests have failed. You get so used to struggling with the build that real issues slip through the cracks too. It's a terrible situation, and fixing that is pretty much first priority.

Finally, @dlombard nailed it with this point:

> I think that before listing the requirements and features, it would be good to determine the problem(s) that KeystoneJS is meant to resolve (What you want Keystone to be ultimately). Right now, Keystone is highly flexible; I could think of it as the swiss knife of web development as it allow to create APIs, web apps and CMSs. To me it makes it an amazing product. However, that flexibility brings its load of difficulties and may result in a pretty long list of requirements and features. At the very least, defining what KeystoneJS should be (or what problem it solves) will help steering future conversations and development in the right direction. It will also serve as a reference point, ensuring that reqs and features will make KeystoneJS more of what it should be.

I couldn't agree more, and will start with that this week.

Regarding *supporting multiple databases (database agnostic) or developing a custom Keystone model like loopback*, I've never wanted to take on building an ORM-style thing in addition to Keystone. We do need _some_ rich model functionality (and probably slightly more than exists right now to decouple from mongoose) but this is where I think GraphQL shines.

---

Ultimately I think the architecture will go something like this:

At the core, you define *Keystone List Schemas*.

Optionally, you can then (one or both)

* (1) Create an generated, extensible Admin UI that is based on those schemas, and expects to run against a compatible back-end API (probably GraphQL)
* (2) Start a node.js app that serves an API based on those schemas, connected to a supported database platform (for use by the Admin UI or your App like graph.cool / firebase / etc)

Additionally, make it easy to (one or more of)

* Serve a node.js server app (w/ express) including session middleware / etc
* Build and serve a front-end app (React, Vue, Angular, etc)

So each part of the above picture would be optional and independent, except for the idea of starting with List Schemas, which has always been the core value that Keystone has provided.

---

Hopefully that gives enough of the picture of where we think this is going for now 🙂 

In the meantime @Noviny and I willl try and get some more plans in place so we can all start tackling the immediate concerns, and will be in touch asap @autoboxer

**Comment by gautamsi** [comment]
@JedWatson  thanks for commenting on TypeScript case, good thing with Typescript is that I can give you pure ES6+ code with not much effort. Effort I put to get it to TypeScript is enough to get ES6+ code out of it. if core idea is to get to ES8, it could also be starting point.

I can post the ES6 version if anyone want to take a look. 

It is good to see the Keystone creators have commitment, I do not have to force TypeScript to it.

**Comment by autoboxer** [comment]
@JedWatson and @Noviny, I really appreciate you bringing me onto the team, and I'm eager to get started, however, until I've spoken to the two of you and feel I've learned your process, I won't feel comfortable commenting on issues or merging pull requests.  I'm excited to help create momentum in this repo again, and while it's going to take me time to get up to speed, there's plenty I'd like to help with.  I can be reached at jared.j.collier@gmail.com, and I'd love to gain access to the Slack channel as well if possible.  I know you're both busy and have a lot to do related to the next release, but please utilize me for anything you may not have time or desire to do.

**Comment by benyto** [comment]
I've been using Keystonejs for some months now, and I'm really impressed by its features and producticivty (thanks Jed and all the team) and would like also to see this project surviving and evolving and I'm willing to help (I'm a developer).

Unfortunalty nowdays most of the "new" frameworks finish by dying just because they can't bring their level to an acceptable entreprise level that can ensure a sufficient financial benificits for people and entreprises backing them. 

IMHO that's should be a priority for this project : bring Keystonejs to an entreprise level. 

For me that means :

- A more powerfull product features :
    * Nested lists and and ORM supporting also RDBMS 
    * Customizable Admin UI
    * Multilingual support
    * A plugin system (compatible with npm). E-commerce and content management system plugins could be the best ones to start with
    * A testing path
    * Support for newest nodejs version

- An up-to-date Keystonejs website where we can see
    * a good documentation
    * product roadmap
    * blog, case studies (how we implemented project X, challenges, why not PHP frameworks or Rails, productivity ...)
    * security compliance/audits
    * deployment (chef, vagrant ...)
    * performance (benchmarks, optimizations, scalling ...)
    * migration path from other cms (wp, drupal ...)
    * partners, training, certifications ...

I believe Keystonejs is closest to those requirements than any other Nodejs framework, or even any new recently trending programming language, this of course will need time/money investment, but I'm sure this can bring Keystonejs (and Thinkmill) to another incommensurable level. Just my 2 cents :-)

**Comment by webteckie** [comment]
I've been out of keystone for a while but before I paused I was pushing a Typescript system for it. I think @gautamsi has taken that to another level and I would greatly encourage any refactoring done to make use of typescript. A lot of the code complications with keystone are due to nasty c-like syntax all over. Typescript will give it more like a c++/c# feel to it, which would make it way easier to reason about things. The type checking will also eliminate a lot of obscure bugs. I know some people have reservations about learning a new language but I think this one is worth learning and last time I used it didn't take a lot of effort to get going. Visual Studio Code has also come a long way and makes for a perfect typescript IDE. Finally, I was checking out cypress for e2e testing and even though I think it's still in beta it is looking very promising.

**Comment by nilsweb** [comment]
Hey everyone,

I was recently searching for a Node.js based CMS for a future project and came across Keystone. After playing with it for a little while, I really started to love it. The whole system seems to be focused on the developer side, making it easy to add custom data structures and functionality. However, there are some parts of Keystone that are more "restricted" on that part (e.g. customizing the Admin UI isn't really supported at the moment).

Therefore, I would like to help with the future development of Keystone as well. First, I thought that the project might be abandoned but luckily, I found this thread telling me otherwise.

Having said all that, if you need _any_ help bringing this project forward, e.g. sorting through GitHub issues/PRs and collecting those problems/ideas/etc. for a future roadmap (as suggested by @JedWatson), you can count me in!

In my opinion, Keystone has great potential becoming a well-established base from where future projects can be built more easily and effectively.

**Comment by chenglx** [comment]
@benyto 's list is very nice one. Especially these features are very attractive:
Nested list, Custom Admin UI, Multilingual support and Test approach

**Comment by webteckie** [comment]
@benyto I'm curious what you mean by "A testing path" in your suggestions. Mind explaining that further?

Also, although @Noviny has been identified as a point of coordination, I think that keystone's as a whole is complicated enough that it should probably be broken down into different areas/aspects and each area/aspect assigned a lead to take charge of the work in it. That lead should then recruit/encourage others to help out and to ensure that area/aspect of keystone is well documented and understood by multiple people since as we know people come and go. A slack channel should be used for all leads to coordinate cross-area/aspect work. An area/aspect should also have its own slack channel where the lead and the people helping with it coordinate all the work. I could think of the following main areas:

1) Overall keystone design/guidelines/technologies
2) Keystone's Admin UI
3) Keystone's Server
4) Keystone's Backend (take note: Microsoft is developing Cosmos DB, which supports multi-models including MongoDB, GraphQL, SQL, etc.  Why can keystone go in that direction as well?)
5) Keystone's Integration Testing (indispensable to ensure the end-to-end paths do not regress)
6) Keystone's Installation/Deployment
7) Keystone's Documentation
8) Keystone's Demo Site(s)
9) Keystone's Issue Management/Triaging/Support
10) Keystone Pull Request and Release Management
11) Keystone's Security/Permissions/Authentication
12) Keystone's Multi-Tenancy (yes, it should be possible for someone to host a site where multiple tenants can coexists with their own storage)
13) etc.

I'm sure some of the above areas are big enough that these themselves probably should be broken down further into different features and managed as such. But the point is that no one person can possibly know this whole system. It should be a coordinated effort for multiple leads.

**Comment by semeano** [comment]
Good to see that this project is not dead. There's nothing like it for mongo. I've forked it and added _List_ and _Object_ field types since I really needed for another project.

You can count me in to help Keystone moving forward.

**Comment by Norbz** [comment]
Im glad to see @JedWatson and @Noviny back in the game.

I am using keystoneJS instensively for it's backend, and I love it. I agree when Jed said that Keystone is agnostic in term of frontend Framework. I Myself coupled it with Nuxt JS for serverside rendering of VueJS powered page. I do not use KeystoneJS for it's template rendering part, but use it for API calls and dynamic JSON mostly.

I found it pretty easy to tweak towards our needs, but I agree that that lack of flexibility of the Admin part is the most frustrating thing.

I am looking forward to seeing the future and Keystone.

Cheers !

**Comment by hades217** [comment]
Hi.

Is Keystonejs still updating? We [Jiangren](https://jiangren.com.au) are using keystone for a while who are looking forward new features

**Comment by morenoh149** [comment]
There's a lot of good will in this thread. @JedWatson should outline next steps.

What is the process for merging small contributions like [this](https://github.com/keystonejs/keystone/pull/4169)?

**Comment by PTZ2002** [comment]
Hi,

I'm developing application with keystone since almost a year, was searching for a good documentation about the exact flow of the keystone framework, a road map what classes are called to run the application.
This would help new and old developers to with the exact understanding of the cms.

@JedWatson I would like to create a plugin system for keystone since for me It would be hard to upgrade a custom keystone to the latest version. If we could decide an idea for this I would do my best to implement it for keystone.

**Comment by autoboxer** [comment]
@JedWatson and @Noviny, please provide an update and next steps.  There is a decent sized community that showed up for this thread, and letting momentum die is what got us here in the first place.  As @morenoh149 stated, there is a lot of good will in this thread and it would be a shame to see nothing come of it.  If you're too busy and you need help, please communicate that to all of us.  I'm busy through the end of the week releasing a project for a client, but I'm hoping to see activity, direction, and a place where I can jump in once I'm done.

**Comment by Noviny** [comment]
Hi,

The most important first next step is to allow things to be moving again. This means we need to have a process for triaging issues and PRs. This is something I really want to call on your help to do, to make sure PRs get reviewed.

On a broader level, we need to make contributing easy. Currently manually testing is done out of a [secondary repository](https://github.com/keystonejs/keystone-test-project) which means validating keystone changes can't be done by just downloading the repository and running it. Combined with the current failing tests means that even small changes take time to verify, which paralyses quick review.

We've been talking about moving to a mono-repo of:

1) keystone core
2) keystone tests
3) the test project
4) the documentation website

Doing so is great because it makes it easy to make and validate changes. It also breaks all existing pull requests, so can't be done until we are confident there are no pull requests that should be merged.

Open discussion on what my role should be: I want to help the community become unblocked from advancing keystone even when Thinkmill is not driving it. Beyond having someone outside of TM who can approve PRs, what do we need to deliver to help people work on keystone?

Going to move management out of this thread on Sunday and into a better tool for it, so we can start to be more granular about what needs to be done. Will post a link to that once it is set up.

@autoboxer I sent you an email. Let's make sure you have a clear path to jump in next week 🙂

**Comment by autoboxer** [comment]
@Noviny, thanks for reaching out.  I'm looking forward to helping wherever it's needed.

**Comment by jacargentina** [comment]
@Noviny waiting for that link. Also want to be on the wagon to help. Cheers all!

**Comment by Noviny** [comment]
My four priorities in approximately the order of importance are:

1. Getting the tests passing again
2. Resolving PRs
2a. Developer experience
3. Getting the updated website publishing to netlify again, finalise the new site
4. Resolving issues.
5. New Features

1 and 3 are big tasks that are going to require concerted effort to resolve. I'm going to be using spectrum to talk about these, and working with @autoboxer to get these resolved. We can't get 4 out of beta until these are done.

2 and 4 are better suited to short bursts, and pickup work. I'm going to be trying to resolve at least one PR an evening. I'm likely going to be working newest to oldest.

2a is cheekily inserted as an aside on 2. This means making a clear documentation for a contributor to replicate and solve their problem int the test app so that a reviewer can easily follow it, as well as looking at better tooling for devving.

5 I want to push to https://github.com/keystonejs/keystone/projects/2 for now while we get 1 and 3 done.

I'm going to be using [spectrum](https://spectrum.chat/keystonejs) as a place to chat about things, rather than slack, as access is easier, and we can start new topics easily when we want to split them.

**Comment by tmvanetten** [comment]
@Noviny I would be willing to contribute covering the cost of slack, set up an easy way to join, and organize slack to prevent all interested parties from having to roll in yet another messaging platform.  If interested in the contribution let me know.

**Comment by Avcajaraville** [comment]
@Noviny Im an experienced FE (6+ years of experience); Im really looking forward for keynote to keep pushing as THE CMS for node. Any channel (slack?) where I can collaborate ? Please, let me know if I can help somehow :)

**Comment by gcortese** [comment]
@Noviny I can have a person in my team help with the work. I would like to see Keystone 4 stable/ documented so that we can base our applications on it. Can we chat to understand how we can collaborate ? Thanks !

**Comment by oeddyo** [comment]
any updates?

**Comment by oeddyo** [comment]
as a new user, it's sad to see the momentum dead again. Would be great some of the main contributors can lead or at least respond to the questions regarding future directions. I understand people have busy schedule, but it's just sad to see an awesome project die :|

**Comment by christroutner** [comment]
A tip for those frustrated by the lack of community. I've been researching
alternatives to KeystoneJS and the only thing that comes close is Strapi:

https://strapi.io/

They have an Admin UI similar to Keystone, but they are in Alpha and things
are still pretty buggy. Still, they have a pretty active community.

I've personally moved away from Express-based web apps (like Keystone) and
adopted Koa. I'm currently using this koa-boilerplate for the new web-apps
I'm building:
https://github.com/adrianObel/koa2-api-boilerplate

Except I've forked it and removed Babel as a dependency:
https://github.com/christroutner/babel-free-koa2-api-boilerplate

The Koa boilerplate is all about simplicity. No overhead of an Admin UI or
other 'features'. It's about starting small, clean, and going from there.
It also lends itself well to test-driven development (TDD).

Food for thought.

**Comment by WingedToaster** [comment]
I’m still messing around with it but Meteor With SimpleSchema, Collections2, and Autoform also seems to be a decent replacement.

https://github.com/meteor/meteor

**Comment by abumalick** [comment]
https://feathersjs.com/ with https://marmelab.com/admin-on-rest/

or https://loopback.io/

**Comment by Norbz** [comment]
@autoboxer @Noviny Did you guys manage to work on recreating some momentum on this ?
I understand that you need some time to make things able to move again, can you share any status update ?

Thanks !

**Comment by morenoh149** [comment]
loopback I've heard has a perverse incentive to not keep their software up to date or provide much of an upgrade path between major versions.

I used meteor 2 years ago on a project and I've heard only improvements have been made since then. It's worth a look. And then perhaps build out an admin panel like https://activeadmin.info/ or https://djangobook.com/django-admin-site/
Then you'd have the fullstack js CMS you'd want.

**Comment by WingedToaster** [comment]
@morenoh149 https://adminlte.io is really nice. Theres a decent meteor package out there that already integrates with it.

**Comment by christroutner** [comment]
I created ConnextCMS (http://connextcms.com/) which combines that AdminLTE
template on top of KeystoneJS.

-Chris Troutner

**Comment by TerrenceRC** [comment]
Anybody have any solid info of if the core group that initially wanted to jumpstart the project back up is still progressing? Or if the interest is stagnant again.

**Comment by morenoh149** [comment]
@TerrenceRC all you see in this thread is all there is. You could ping all participants of this thread and try to contribute to a keystone fork or some other effort. Fundamentally thinkmill has retained commit access to keystone. It's completely within their right to do so but have not had the bandwidth to push the project forward themselves.

**Comment by danielmahon** [comment]
I'm still using KeystoneJS for a few small production sites but the recent lull in activity has gotten me to discover some of the newer "Headless" CMSs. I'm actually pretty impressed with the idea of using a static-site generator like [GatsbyJS](https://www.gatsbyjs.org/) and being able to pick and choose the CMS that works best for a project (paid/open source). Particularly [GraphCMS](https://graphcms.com/) and [NetlifyCMS](https://www.netlifycms.org/). I also like the approach since development more closely mirrors how I'm currently working on ReactJS mobile/native apps. For anyone interested, I created a simple starter/boilerplate for integrating GatsbyJS with GraphCMS or NetlifyCMS. Check it out if you want, I will probably be converting a couple smaller Keystone sites to this as a test.

GraphCMS version with in-place editing (master branch)
https://github.com/danielmahon/gatsby-starter-procyon 

NetlifyCMS git-based CMS (netlifycms branch)
https://github.com/danielmahon/gatsby-starter-procyon/tree/netlifycms

**Comment by JedWatson** [comment]
Clearly another update from me is overdue. For the record, @morenoh149 isn't speaking for the project and we haven't abandoned Keystone.

After spending a couple of months trying to pick up the roadmap, we've realised that it's not really working - there is so much technical debt in the current state of the project that we're spending more time churning than actually making progress.

Also, to be honest, Keystone <= 4 was written for how we wanted to build web sites and apps in node.js back in 2014, and things have come a long way since then.

Having accepted that, we're splitting into two efforts: one will simply clean up the last release of 4.0, remove the `beta` tag and publish it so that the current version can live and get patches as necessary. The main final blocker to that is getting the docs ready to be published (specifically, they're running an old version of Gatsby which won't build right now)

The second effort is to re-implement Keystone in a completely modern architecture. Including a using a monorepo, ES8 (Node 8+), webpack, React 16, GraphQL and Apollo. We're tackling the problems that were previously really hard to solve (like custom field types, etc) up front so we don't end up with the same issues down the track.

I'm not clear how we'll deal with the cut-over to the next version yet; trying to hold on to backwards compatibility has been a mistake (that I take responsibility for) and has seriously hurt the project. Our focus right now is to get rid of everything that's holding the project back and making it hard for people to contribute, then we'll deal with the upgrade path when we know what the future looks like.

For the record, if there's something else out there that works for you, please - use the best tool(s) that you can find for whatever you're doing. This is open source, not a competition 🙂

For me, I am feeling really good about this reboot. I honestly haven't enjoyed working on Keystone this much in years, and I look forward to sharing it with anyone who's still interested when it's ready.

**Comment by ttsirkia** [comment]
Thanks @JedWatson for your thoughts!

In my opinion, using minimum effort to publish 4.0 with up-to-date dependecies would be the best choice in the near future so that it is possible to update old KeystoneJS 0.3 projects which run only with ancient Node versions. Then, a clear announcement that no major feature updates or changes are coming to 4.0 and the next version might be completely rewritten at some point when the roadmap is clear.

The version 4.0 has been waiting the final release so long that it starts becoming obsolate sooner or later.

**Comment by BinaryBen** [comment]
While @morenoh149 might not represent the core team, I'd say the sentiments are felt by the rest of the non-core team. 

It's easy to keep popping in with a message once in a blue moon that things are progressing behind the scenes. In your teams defence, they probably are. But you have open source developers ready to help code or promote the project, and no way for them to do so. For all purposes, this project looks dead. 

If you're rebuilding from scratch, cut your losses and put 100% of the team on that. With the speed that tech moves, whatever you publish at this point is going to be obsolete anyway. Also, prioritise the pathway for community involvement above all else. 

It may not be a competition in your eyes, but open source projects need a decent mind share to stay healthy.

**Comment by autoboxer** [comment]
I've been mostly silent from when I joined the Keystone team until now as I was trying to work out exactly where I would be helping with the project, and didn't want to overstep my bounds early on.  It's clear that communication needs to improve on this project, and the frustration from the community is real and should be addressed.  I started this issue back in January based on the frustration I felt, and the fact that I had projects reliant on Keystone but wasn't sure about its longevity.  Based on wanting to address all of our frustrations but not being able to, I've requested that on top of helping with code, Jed allow me to act as a second community manager, and he was willing to let me take on the role.  I'm hoping this message acts as the catalyst for strong community involvement and transparency that Keystone would benefit from moving towards, and I'm going to do my best with the limited time I have to help make that happen.  As a first step, I've been made an admin of the Slack channel, and want to move the conversation there to shorten the response time and start a dialog about what's needed.  I'm listing my personal email below for everyone here, and everyone who reads this issue in the future, please use it to request access if you don't already have it.  I'd like to spend my time as a community manager bringing everyone together to discuss everything from how people can help to what we need to do to get useful information into your hands, as well as freeing up time for the folks at Thinkmill to focus on code and gathering community feedback and requests for them.  I've never taken on a role like this and am excited to help Keystone thrive moving forward, but will need all of your help figuring out the best way to do that.  Hopefully I'll see you on Slack and we can take the next step together.

for invites to Slack, email me at jared.j.collier@gmail.com

**Comment by 1337cookie** [comment]
Graph.cool is really awesome. So are we going to have a back-backend for making our schemas/relationships like graphcool and then a backend for editing the data? and can takeCRUD components from the backend into the frontend with react? Shutup and take my money Jed!

**Comment by stennie** [comment]
As mentioned in the follow-up discussion on #4638 (aka "What's the state of Keystone (again)"), more contributors have recently been onboarded (👋) and we're working to get the project to a healthy state with more timely & effective communication, additional contributors, and a Keystone 4 release without a beta tag.

There's now an open invite for anyone interested to [join the KeystoneJS Community Slack](https://launchpass.com/keystonejs) for feedback and discussion. Join the KeystoneJS Slack and introduce yourself!

Regards,
Stennie

**Comment by eddr** [comment]
Is there a codebase for the reboot?
